### PR TITLE
OKRS-24-55

### PR DESCRIPTION
**Pull request by dmytrotsko**
**Prerequisites**:

- [x] Branch is up-to-date with the branch to be merged with, i.e. `dev`
- [x] Build is successful
- [x] Code is cleaned up and formatted

### Summary

"latest known data" link now preserves context.
Set default date to the date when we have data for all 3 sensors in 'Overview'. When date is present in URL, the date doesn't change to "default date". Added "NoRecentDataWarning" in order to nofity user about "old" data

### CODE REVIEW

File: src/blocks/NoRecentDataWarning.svelte [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  /**
+   * @type {import('../stores/params').SensorParam}
+   */
+  export let deathSensor;
+  /**
+   * @type {import('../stores/params').SensorParam}
+   */
+  export let hospitalAdmissionSensor;
+
+  let todaysDate = new Date();
```
Comment (melange396) [inline src/blocks/NoRecentDataWarning.svelte:24]: do we need `todaysDate`?  can we use `date` instead?  i think theyre set to the same value, right?

**Review by melange396 — CHANGES_REQUESTED** [review]
This looks great!  I especially appreciate how you got rid of the asterisks and added tooltip hovers for the latest dates of each of the 3 highlighted indicators.

There is one behavior that we might want to change though (and we can merge this PR and fix it later, if necessary) -- when you go to the dashboard without any url parameters (like https://deploy-preview-1243--cmu-delphi-covidcast.netlify.app/ ), it shows the warning message as appropriate.  However, when you go to the dashboard with a date specified that matches `minMaxDate`, it still shows the warning message (for instance, as of today `20240416`, the most recent date with all 3 indicators available is `20240329`, so today this happens when you visit https://deploy-preview-1243--cmu-delphi-covidcast.netlify.app/?date=20240329 ).  This doesn't give ***incorrect*** information, but it could be fairly confusing for users.

File: src/modes/summary/Overview.svelte [hunk 2]
```diff
@@ -27,33 +25,55 @@
...
-        return t;
-      }
-      return fetcher.fetch1Sensor1Region1DateTrend(sensor, region, DateParam.box(sensor.timeFrame.max));
-    });
-  }
-
   $: trends = fetcher.fetchNSensors1Region1DateTrend([CASES, HOSPITAL_ADMISSION, DEATHS], region, date);
   $: casesTrend = trends[0];
-  $: hospitalTrend = fetchFallback(fetcher, HOSPITAL_ADMISSION, region, trends[1]);
+  $: hospitalTrend = trends[1];
```
[the remark this answers is not in the document]
Comment (melange396) [reply src/modes/summary/Overview.svelte:30]: yeah, after some more looking, i think its fine to get rid of the fallback stuff here.

File: src/modes/summary/Overview.svelte [hunk 3]
```diff
@@ -1,14 +1,12 @@
...
   import KPIValue from '../../components/KPIValue.svelte';
   import KPIChange from '../../components/KPIChange.svelte';
-  import { DateParam, SensorParam } from '../../stores/params';
-  import { formatDateDayOfWeek } from '../../formats';
+  import { SensorParam } from '../../stores/params';
   import SensorUnit from '../../components/SensorUnit.svelte';
-  import IndicatorAnnotations from '../../components/IndicatorAnnotations.svelte';
   import MaxDateHint from '../../blocks/MaxDateHint.svelte';
-  import IndicatorWarning from '../../blocks/IndicatorWarning.svelte';
+  import NoRecentDataWarning from '../../blocks/NoRecentDataWarning.svelte';
```
[the remark this answers is not in the document]
Comment (melange396) [reply src/modes/summary/Overview.svelte:7]: It looks like you have already taken care of replacing the important parts of IndicatorAnnotations and IndicatorFallbackWarning, so i think we are good!

File: src/blocks/IndicatorWarning.svelte [hunk 4]
```diff
@@ -38,8 +38,9 @@
       <div data-uk-alert class="uk-alert-warning">
         The indicator "{sensor.name}" is not available for {formatDateYearDayOfWeekAbbr(date.value)}, yet. The latest
         known data is available on
-        <a href="?date={formatAPITime(sensor.timeFrame.max)}" on:click={switchDate}
-          >{formatDateYearDayOfWeekAbbr(sensor.timeFrame.max)}</a
+        <a
+          href="?date={formatAPITime(sensor.timeFrame.max)}&{window.location.search.split('&').slice(1).join('&')}"
```
[the remark this answers is not in the document]
Comment (dmytrotsko) [reply src/blocks/IndicatorWarning.svelte:42]: I didn't mean to replace "IndicatorWarning.svelte" by "NoRecentDataWarning.svelte", but I didn't want to change existing "Indicator.svelte" either.
I would prefer to add comment and leave both of them (at least for now).

**Comment by dmytrotsko** [comment]
> This looks great! I especially appreciate how you got rid of the asterisks and added tooltip hovers for the latest dates of each of the 3 highlighted indicators.
> 
> There is one behavior that we might want to change though (and we can merge this PR and fix it later, if necessary) -- when you go to the dashboard without any url parameters (like https://deploy-preview-1243--cmu-delphi-covidcast.netlify.app/ ), it shows the warning message as appropriate. However, when you go to the dashboard with a date specified that matches `minMaxDate`, it still shows the warning message (for instance, as of today `20240416`, the most recent date with all 3 indicators available is `20240329`, so today this happens when you visit https://deploy-preview-1243--cmu-delphi-covidcast.netlify.app/?date=20240329 ). This doesn't give _**incorrect**_ information, but it could be fairly confusing for users.

Yeah, I agree that it is uncorrect behaviour and I wanted to fix that, but I have not found how to do that yet.

File: src/blocks/NoRecentDataWarning.svelte (hunk 1 above)
Comment (dmytrotsko) [reply src/blocks/NoRecentDataWarning.svelte:24]: It is a tricky way as `date` is stored in some kind of "storage" and it is global, so when date value is changed from code/web page/url, it is reflected in that variable's value.
So I had to create one more variable to get today's date.

File: src/modes/summary/Overview.svelte [hunk 5]
```diff
@@ -27,33 +25,56 @@
...
+  let warningType = 1;
+  beforeUpdate(() => {
+    if (date.value > minMaxDate) {
+      warningType = 2;
+    } else if (date.value < minMaxDate) {
+      warningType = 0;
+    } else {
+      warningType = 1;
+    }
+  });
```
Comment (melange396) [inline src/modes/summary/Overview.svelte:67]: can we do the `date.value != minMaxDate` comparisons inside `NoRecentDataWarning.svelte`?  That way we wouldn't have to pass `warningType`...  Or do we need to do the comparisons inside of the `beforeUpdate()` event here?

File: src/blocks/NoRecentDataWarning.svelte [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  function switchDate() {
+    date.set(minMaxDate);
+  }
+</script>
+
+{#if warningType === 1}
+  <div data-uk-alert class="uk-alert-warning">
+    <p>
+      This date {formatDateYearDayOfWeekAbbr(minMaxDate)} is the most recent that has data for all three of the highlighted
```
Comment (melange396) [inline src/blocks/NoRecentDataWarning.svelte:23]:
```suggestion
      This date, {formatDateYearDayOfWeekAbbr(minMaxDate)}, is the most recent that has data for all three of the highlighted
```

File: src/blocks/NoRecentDataWarning.svelte [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      indicators. You can mouse over the tooltips just below this message to see the latest available date for each indicator.
+      Note that the latest available date may be different for each indicator.
+    </p>
+  </div>
+{/if}
+
+{#if warningType === 2}
+  <div data-uk-alert class="uk-alert-warning">
+    <p>
+      This date {formatDateYearDayOfWeekAbbr(date.value)} does not yet have data for all of the highlighted indicators.
```
Comment (melange396) [inline src/blocks/NoRecentDataWarning.svelte:33]:
```suggestion
      This date ({formatDateYearDayOfWeekAbbr(date.value)}) does not yet have data for all of the highlighted indicators.
```

**Review by melange396 — CHANGES_REQUESTED** [review]
This is great!  Just two little things for punctuation, and then one comment/question about code location.

File: src/modes/summary/Overview.svelte (hunk 5 above)
Comment (melange396) [reply src/modes/summary/Overview.svelte:67]: i talked to Dmytro on slack, he is going to look at this but we will release the current state in the meantime as the functionality is doing what its supposed to.

**Review by melange396 — APPROVED** [review]
awesome!

**Comment by melange396** [comment]
thanks Dmytro, youre the best!

**Pull request merged**